exec-service-wasmer/src/wasmer_metering.rs

```diff
@@ -0,0 +1,189 @@
+use crate::get_opcode_cost;
+use elrond_exec_service::OpcodeCost;
+use loupe::{MemoryUsage, MemoryUsageTracker};
+use std::cell::RefCell;
+use std::mem;
+use std::rc::Rc;
+use wasmer::wasmparser::{Operator, Type as WpType, TypeOrFuncType as WpTypeOrFuncType};
+use wasmer::{
+    ExportIndex, FunctionMiddleware, GlobalInit, GlobalType, Instance, LocalFunctionIndex,
+    MiddlewareError, MiddlewareReaderState, ModuleMiddleware, Mutability, Type,
+};
+use wasmer_types::{GlobalIndex, ModuleInfo};
+
+const METERING_POINTS_LIMIT: &str = "metering_points_limit";
+const METERING_POINTS_USED: &str = "metering_points_used";
+
+#[derive(Clone, Debug, MemoryUsage)]
+struct MeteringGlobalIndexes(GlobalIndex, GlobalIndex);
+
+impl MeteringGlobalIndexes {
+    fn points_limit(&self) -> GlobalIndex {
+        self.0
+    }
+
+    fn points_used(&self) -> GlobalIndex {
+        self.1
+    }
+}
+
+#[derive(Debug)]
+pub struct Metering {
+    points_limit: u64,
+    opcode_cost: Rc<OpcodeCost>,
+    global_indexes: Rc<RefCell<Option<MeteringGlobalIndexes>>>,
+}
+
+impl Metering {
+    pub fn new(points_limit: u64, opcode_cost: Rc<OpcodeCost>) -> Self {
+        Self {
+            points_limit,
+            opcode_cost,
+            global_indexes: Rc::new(RefCell::from(None)),
+        }
+    }
+}
+
+unsafe impl Send for Metering {}
+unsafe impl Sync for Metering {}
+
+impl MemoryUsage for Metering {
+    fn size_of_val(&self, tracker: &mut dyn MemoryUsageTracker) -> usize {
+        mem::size_of_val(self) + self.global_indexes.size_of_val(tracker)
+            - mem::size_of_val(&self.global_indexes)
+    }
+}
+
+impl ModuleMiddleware for Metering {
+    fn generate_function_middleware(
+        &self,
+        _local_function_index: LocalFunctionIndex,
+    ) -> Box<dyn FunctionMiddleware> {
+        Box::new(FunctionMetering {
+            accumulated_cost: Default::default(),
+            opcode_cost: self.opcode_cost.clone(),
+            global_indexes: self.global_indexes.borrow().clone().unwrap(),
+        })
+    }
+
+    fn transform_module_info(&self, module_info: &mut ModuleInfo) {
+        let points_limit_global_index = module_info
+            .globals
```

Comment:
This has nothing to do with WASM globals, right?

Reply:
Unfortunately for us, it has. TBD